### PR TITLE
3.0.0 main feature

### PR DESCRIPTION
**Pull request by ota42y**
- [x] drop old ruby version #146
- [x] refactoring #149
- [x] OpenAPI3 #145 
- [x] do other tasks
  - see https://github.com/interagent/committee/issues/126#issuecomment-448188000
- [x] beta release
- [x] try to use beta version

### CODE REVIEW

File: test/middleware/request_validation_open_api_3_test.rb [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    params = {
+        "string_post_1" => "cloudnasium"
+    }
+    header "Content-Type", "application/json"
+    post "/characters", JSON.generate(params)
+
+    assert_equal 200, last_response.status
+  end
+
+  it "not parameter requset" do
```
Comment (brandur) [inline test/middleware/request_validation_open_api_3_test.rb]: Minor spelling: *request

File: CHANGELOG [hunk 2]
```diff
@@ -1,4 +1,20 @@
...
+	Drop old ruby version (#146)
+
+2.4.0 xxxxx
+	Add methods for harmless upgrade (#199, #195)
+	Support filepath option (#191)
+	Change change validate_errors option to validate_success_only (#187)
+	Bugfix for coerce_recursive option in request params (#162)
+	Path definition fewer matches in OpenAPI2 (#160)
+	Deprecated test method (#157)
+	Add `error_handler` option (#152)
```
Comment (brandur) [inline CHANGELOG]: Sorry, I rewrote some of the 2.4.0 CHANGELOG entries in 114edc73bc433814fd25d2dea10b4476b48500db so you may have to rebase. I would just replace this whole block with the new set.

File: lib/committee/bin/committee_stub.rb [hunk 3]
```diff
@@ -8,6 +10,8 @@ class CommitteeStub
       def get_app(schema, options)
         cache = {}
 
+        raise Committee::NotSupportOpenAPI3.new("OpenAPI3 not support stub") unless schema.support_stub?
```
Comment (brandur) [inline lib/committee/bin/committee_stub.rb]: Could you rename:

* `NotSupportOpenAPI3` -> `OpenAPI3Unsupported`
* "OpenAPI3 not support stub" -> "Stubs are not yet supported for OpenAPI 3"

File: lib/committee/drivers.rb [hunk 4]
```diff
@@ -17,14 +19,39 @@ def self.driver_from_name(name)
...
+    # @param [String] filepath
+    # @return [Committee::Driver]
+    def self.load_from_file(filepath)
+      case File.extname(filepath)
+      when '.json'
+        load_from_json(filepath)
+      when '.yaml', '.yml'
+        load_from_yaml(filepath)
+      else
+        raise "committee filepath option support '.yaml', '.yml', '.json' files only"
```
Comment (brandur) [inline lib/committee/drivers.rb]: Could you reword this to:

"Committee only supports the following file extensions: '.json', '.yaml', '.yml'"

File: lib/committee/drivers.rb [hunk 5]
```diff
@@ -85,6 +105,15 @@ class Schema
       def driver
         raise "needs implementation"
       end
+
+      def build_router(options)
+        raise "needs implementation"
+      end
+
+      # OpenAPI3 not support stub but JSON Hyper-Schema and OpenAPI2 support
+      def support_stub?
```
Comment (brandur) [inline lib/committee/drivers.rb]: Could we rename this method to `supports_stub?`?

File: lib/committee/drivers.rb [hunk 6]
```diff
@@ -85,6 +105,15 @@ class Schema
       def driver
         raise "needs implementation"
       end
+
+      def build_router(options)
+        raise "needs implementation"
+      end
+
+      # OpenAPI3 not support stub but JSON Hyper-Schema and OpenAPI2 support
```
Comment (brandur) [inline lib/committee/drivers.rb]: Can you reword this comment to:

> Stubs are supported in JSON Hyper-Schema and OpenAPI 2, but not yet in OpenAPI 3

File: lib/committee/drivers/open_api_2.rb [hunk 7]
```diff
@@ -17,6 +17,14 @@ def default_query_params
       true
     end
 
+    def default_coerce_date_times
+      false
+    end
```
Comment (brandur) [inline lib/committee/drivers/open_api_2.rb]: Would you mind trying to see if we can keep all these methods ordered alphabetically in each file? I think that will just involve moving `default_coerce_date_times` to the top position above `default_coerce_form_params`.

File: lib/committee/middleware/base.rb [hunk 8]
```diff
@@ -24,70 +21,29 @@ def call(env)
...
-      elsif schema.is_a?(Hash)
-        warn_hash_deprecated
-      elsif schema.is_a?(JsonSchema::Schema)
-        warn_json_schema_deprecated
-      end
+    class << self
+      def get_schema(options)
+        schema = options[:schema]
+        unless schema
+          schema = Committee::Drivers::load_from_file(options[:filepath]) if options[:filepath]
```
Comment (brandur) [inline lib/committee/middleware/base.rb]: As mentioned in #203, let's try to do a bulk rename from `filepath` to `schema_path`.

File: lib/committee/middleware/stub.rb [hunk 9]
```diff
@@ -1,3 +1,5 @@
+# Don't Support OpenAPI3
```
Comment (brandur) [inline lib/committee/middleware/stub.rb]: Change to:

> `Stub` is not yet supported in OpenAPI 3

File: lib/committee/middleware/stub.rb [hunk 10]
```diff
@@ -12,6 +14,8 @@ def initialize(app, options={})
       @cache = options[:cache]
 
       @call = options[:call]
+
+      raise Committee::NotSupportOpenAPI3.new("OpenAPI3 not support stub") unless @schema.support_stub?
```
Comment (brandur) [inline lib/committee/middleware/stub.rb]: Change to:

> "Stubs are not yet supported for OpenAPI 3"

File: test/bin/committee_stub_test.rb [hunk 11]
```diff
@@ -25,6 +25,12 @@
     assert_equal true, options[:tolerant]
     assert_equal "1234", options[:port]
   end
+
+  it "OpenAPI3 not support Stub" do
```
Comment (brandur) [inline test/bin/committee_stub_test.rb]: Change to:

``` ruby
it "is not supported in OpenAPI 3" do
```
```

File: lib/committee/schema_validator/open_api_3.rb [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    def initialize(router, request, validator_option)
+      @router = router
+      @request = request
+      @operation_object = router.operation_object(request)
+      @validator_option = validator_option
+    end
+
+    def request_validate(request)
+      path_params = validator_option.coerce_path_params ? coerce_path_params : {}
+      # coerce_query_params(request) if validator_option.coerce_query_params
```
Comment (brandur) [inline lib/committee/schema_validator/open_api_3.rb]: Looks like this isn't needed anymore.

File: lib/committee/schema_validator/open_api_3.rb [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+          new(@operation_object, validator_option).
+          call(status, headers, data, strict)
+    end
+
+    def link_exist?
+      !@operation_object.nil?
+    end
+
+    def coerce_form_params(_parameter)
+      # nothing because when request_schema_validation, check and coerce
```
Comment (brandur) [inline lib/committee/schema_validator/open_api_3.rb]: Could this be reworded to?

> Empty because `request_schema_validation` checks and coerces

(If that's accurate?)

File: test/data/openapi3/normal.yml [hunk 14]
```diff
@@ -0,0 +1,539 @@
+openapi: 3.0.0
```
Comment (brandur) [inline test/data/openapi3/normal.yml]: Quite minor, but would you mind changing these to use the `.yaml` extension instead of `.yml`? This is [the extension recommended by the project](https://yaml.org/faq.html).

File: test/drivers/open_api_3_test.rb [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  end
+
+  it "defaults to query parameters" do
+    assert_equal true, @driver.default_query_params
+  end
+
+  describe "Schema" do
+    describe "#operation_object" do
+      describe "path template" do
+        it "get normal path" do
```
Comment (brandur) [inline test/drivers/open_api_3_test.rb]: Can we change this and all instances below of `it "get ...` to `it "gets a ...`?

File: test/drivers/open_api_3_test.rb [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+          assert_equal({"template_name"=>"test", "nested_parameter" => "abc"}, obj.path_params)
+        end
+
+        it "get twice nested concrete path" do
+          obj = open_api_3_schema.operation_object("/path_template_test/test/abc/finish", "get")
+          assert_equal "/path_template_test/{template_name}/{nested_parameter}/finish", obj.original_path
+          assert_equal({"template_name"=>"test", "nested_parameter" => "abc"}, obj.path_params)
+        end
+
+        it "get ambiguous path" do
```
Comment (brandur) [inline test/drivers/open_api_3_test.rb]: All except this one which should be "gets an ambiguous path".

File: test/drivers_test.rb [hunk 17]
```diff
@@ -32,6 +33,19 @@
       assert_kind_of Committee::Drivers::Schema, s
       assert_kind_of Committee::Drivers::HyperSchema::Schema, s
     end
+
+    it 'load OpenAPI3' do
```
Comment (brandur) [inline test/drivers_test.rb]: Let's say:

``` ruby
it 'loads OpenAPI 3' do
```

File: test/drivers_test.rb [hunk 18]
```diff
@@ -32,6 +33,19 @@
...
+      s = Committee::Drivers.load_from_file(open_api_3_filepath)
+      assert_kind_of Committee::Drivers::Schema, s
+      assert_kind_of Committee::Drivers::OpenAPI3::Schema, s
+    end
+
+    it 'load unsupported file extension' do
```
Comment (brandur) [inline test/drivers_test.rb]: Let's say:

``` ruby
it "errors on an unsupported file extension" do
```

File: test/middleware/base_test.rb [hunk 19]
```diff
@@ -19,43 +19,65 @@ def app
     assert_equal 200, last_response.status
   end
 
-  it "accepts schema string (legacy behavior)" do
-    mock(Committee).warn_deprecated.with_any_args
+  it "accepts just a OpenAPI3 schema object" do
```
Comment (brandur) [inline test/middleware/base_test.rb]: Let's say:

``` ruby
it "accepts an OpenAPI 3 schema object" do
```

File: test/middleware/base_test.rb [hunk 20]
```diff
@@ -19,43 +19,65 @@ def app
...
+    @app = new_rack_app(schema: open_api_3_schema)
+    params = {
+        "name" => "cloudnasium"
+    }
+    header "Content-Type", "application/json"
+    post "/apps", JSON.generate(params)
+    assert_equal 200, last_response.status
+  end
+
+  it "don't accepts schema string" do
```
Comment (brandur) [inline test/middleware/base_test.rb]: Let's say:

``` ruby
it "doesn't accept a schema string" do
```

File: test/middleware/base_test.rb [hunk 21]
```diff
@@ -19,43 +19,65 @@ def app
...
+      post "/apps", JSON.generate(params)
+    end
+
+    assert_equal "Committee: schema expected to be an instance " +
+                     "of Committee::Drivers::Schema.", e.message
   end
 
-  it "accepts schema hash (legacy behavior)" do
-    mock(Committee).warn_deprecated.with_any_args
+  it "don't accepts schema hash" do
```
Comment (brandur) [inline test/middleware/base_test.rb]: Likewise, let's say:

``` ruby
it "doesn't accept a schema hash" do
```

File: test/middleware/base_test.rb [hunk 22]
```diff
@@ -19,43 +19,65 @@ def app
...
+      post "/apps", JSON.generate(params)
+    end
+
+    assert_equal "Committee: schema expected to be an instance " +
+                     "of Committee::Drivers::Schema.", e.message
   end
 
-  it "accepts schema JsonSchema::Schema object (legacy behavior)" do
-    mock(Committee).warn_deprecated.with_any_args
+  it "don't accepts schema JsonSchema::Schema object" do
```
Comment (brandur) [inline test/middleware/base_test.rb]: Likewise, let's say:

``` ruby
it "doesn't accept a JsonSchema::Schema object" do
```

File: test/middleware/base_test.rb [hunk 23]
```diff
@@ -68,6 +90,15 @@ def app
     assert_equal "Committee: schema expected to be an instance of Committee::Drivers::Schema.", e.message
   end
 
+  it "schema not exist" do
```
Comment (brandur) [inline test/middleware/base_test.rb]: Let's say:

``` ruby
it "errors when no schema is specified" do
```

File: test/request_unpacker_test.rb [hunk 24]
```diff
@@ -94,46 +94,80 @@
...
         coerce_form_params: true,
-        schema: schema
+        schema_validator: validator,
       ).call
       assert_equal({ "x" => 1 }, params)
     end
   end
 
-  it "coerces form params with coerce_form_params, coerce_recursive and a schema" do
+  it "coerces form params with coerce_form_params and a OpenAPI3 schema" do
```
Comment (brandur) [inline test/request_unpacker_test.rb]: "an OpenAPI 3 schema"

File: CHANGELOG [hunk 25]
```diff
@@ -1,4 +1,20 @@
 Unreleased
+	Support full committee options in OpenAPI3
+	Fix bug when non defined link with form content type and coerce option (#173)
+	Support check_content_type option for OpenAPI3 (#174)
+
+3.0.0.alpha 2018-12-26
+	OpenAPI 3.0 support (#164)
+	Drop old ruby version (#146)
```
Comment (brandur) [inline CHANGELOG]: Can we say "Drops support for old versions of Ruby"?

**Comment by yhirano55** [comment]
@brandur 

> sorry for being the worst maintainer of all time.

Though you might be not using committee gem, we know you maintain this repository.
We appreciate to your maintainer's work, and we recognize you're so nice maintainer 👏👏

**Comment by brandur** [comment]
Thank you @yhirano55!! Those are very kind words :)

**Comment by ota42y** [comment]
I created 2-x branch.
So we can merge this branch to master.
https://github.com/interagent/committee/tree/2-x
